File: Problems/Leetcode/3001-4000/3201-3300/3275.py
from heapq import heappop, heappush
from sortedcontainers import SortedList
from typing import List

fabs = lambda x: x if x >= 0 else -x


class Solution:
    def resultsArray(self, queries: List[List[int]], k: int) -> List[int]:
        ans = [-1] * len(queries)
        h = []
        for i, (u, v) in enumerate(queries):
            heappush(h, -fabs(u) - fabs(v))  # 最大堆
            if len(h) > k:
                heappop(h)
            if len(h) == k:
                ans[i] = -h[0]
        return ans

        """
        有序列表
        """
        ans = []
        l = SortedList()
        for u, v in queries:
            l.add(fabs(u) + fabs(v))
            ans.append(l[k - 1] if len(l) >= k else -1)
        return ans

File: Problems/Leetcode/3001-4000/3201-3300/3275.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

class Solution {
public:
    vector<int> resultsArray(vector<vector<int>>& queries, int k) {
        int n = queries.size();
        vector<int> ans(n, -1);
        priority_queue<int> pq;
        for (int i = 0; i < n; ++i) {
            pq.emplace(abs(queries[i][0]) + abs(queries[i][1]));
            if (pq.size() > k) {
                pq.pop();
            }
            if (pq.size() == k) {
                ans[i] = pq.top();
            }
        }
        return ans;
    }
};

int main() {
    Solution sol;

    return 0;
}